qutip/wigner.py

```diff
@@ -548,83 +548,297 @@ def _wig_laguerre_val(L, x, c):
+        self._alpha_matrix = _qfunc_alpha_matrix(xs, ys, g)
+        # 16 bytes per complex, 1024**2 bytes per MB.
+        self._size_mb = self._alpha_matrix.conj.size * 16 / (1024**2)
+        self._memory_mb = memory
+        self._max_size = int(self._memory_mb // self._size_mb)
+        self._current_size = 0
+        self._cache = None
+
+    def _alphas(self, size: int):
+        r"""
+        Retrive the full tensor of (the conjugate of) coherent states over all
+        values of :math:`\alpha`, for states of dimension ``size``.
+        """
+        if self._current_size >= size:
+            return self._cache[:, :, :size]
+        if size > self._max_size:
+            requirement = self._size_mb * size
+            raise MemoryError(
+                f"Refusing to precompute up to {size} basis states."
+                f" This would require {requirement:.2f} MB,"
+                f" but only {self._memory_mb} MB is allowed."
+            )
+        if self._cache is None:
+            self._cache = self._alpha_matrix(self._current_size, size)
+        else:
+            self._cache = np.dstack([
+                self._cache,
+                self._alpha_matrix(self._current_size, size)
+            ])
+        self._current_size = size
+        return self._cache
+
+    def _single(self, vector: np.ndarray, alphas: np.ndarray):
+        r"""
+        Get the Q function (without the :math:`\pi` scaling factor) of a single
+        state vector.
+        """
+        return np.abs(np.dot(alphas, (self._g*0.5)*vector))**2
+
+    def __call__(self, state: Qobj):
+        """
+        Get the Husimi-Q function for the given state vector or density matrix,
+        over the coordinates used to initialise the class.  If called multiple
+        times, the states do not need to have the same dimensions, but none of
+        them can have tensor-product structure.
+        """
+        state = _qfunc_check_state(state)
+        alphas = self._alphas(state.shape[0])
+        if state.isket:
+            return self._single(state.full().ravel(), alphas) / np.pi
+        # We don't use Qobj.eigenstates() to avoid building many unnecessary
+        # CSR versions of dense matrices.
+        values, vectors = eigh(state.full())
+        vectors = vectors.T
+        out = values[0] * self._single(vectors[0], alphas)
+        for value, vector in zip(values[1:], vectors[1:]):
+            out += value * self._single(vector, alphas)
+        return out / np.pi
+
+
+def _qfunc_iterative_single(
+        vector: np.ndarray,
+        alpha: _qfunc_alpha_matrix,
+        g: float,
+):
+    r"""
+    Get the Q function (without the :math:`\pi` scaling factor) of a single
+    state vector, using the iterative algorithm which recomputes the powers of
+    the coherent-state matrix.
+    """
+    ns = np.arange(vector.shape[0])
+    out = np.polyval(
+        (0.5*g * vector / np.sqrt(scipy.special.factorial(ns)))[::-1],
+        alpha.conj,
```

Comment:
A comment here saying:

"""
alpha is an instance of the `_qfunc_alpha_matrix` with alpha.conj specifying the grid of 
complex numbers representing points in the phase space.
"""

It was a bit confusing to use `alpha` since I am expecting something like a complex number here